### PR TITLE
surfer: init at v0.2.0

### LINKED ISSUE #299458: Package request: surfer
**Project description**
A waveform viewer with a focus on a snappy usable interface, and extensibility.

**Metadata**

* homepage URL: https://surfer-project.org/
* source URL: https://gitlab.com/surfer-project/surfer
* license: eupl12
* platforms: linux

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by hakan-demirli**
## Description of changes
Supersedes https://github.com/NixOS/nixpkgs/pull/320993

Closes https://github.com/NixOS/nixpkgs/issues/299458.
@Guanran928 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by hakan-demirli** [comment]
Result of `nixpkgs-review pr 324637` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>surfer</li>
  </ul>
</details>

File: pkgs/by-name/su/surfer/package.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  rustPlatform,
+  pkg-config,
+  openssl,
+  wayland,
+  libxkbcommon,
+  libGL,
+}:
+rustPlatform.buildRustPackage rec {
+  pname = "surfer";
+  version = "v0.2.0";
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:13]:
```suggestion
  version = "0.2.0";
```

from the [contributing guidelines](https://github.com/NixOS/nixpkgs/tree/665394853091a08d679eb0c166900cb4f57a5bf5/pkgs#package-naming)

> There are a few naming guidelines:
>
> - The pname attribute should be identical to the upstream package name.
>
> - The pname and the version attribute must not contain uppercase letters — e.g., "mplayer" instead of "MPlayer".
>
> - ***The version attribute must start with a digit e.g., "0.3.1rc2".***

File: pkgs/by-name/su/surfer/package.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+  src = fetchFromGitLab {
+    owner = "surfer-project";
+    repo = pname;
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:17]:
```suggestion
    repo = "surfer";
```
it's bad practice to conflate the package name and repository name. see https://github.com/NixOS/nixpkgs/issues/277994

File: pkgs/by-name/su/surfer/package.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    rev = version;
+    hash = "sha256-C5jyWLs7fdEn2oW5BORZYazQwjXNxf8ketYFwlVkHpA";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [pkg-config];
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:23]:
```suggestion
  nativeBuildInputs = [ pkg-config ];
```

File: pkgs/by-name/su/surfer/package.nix [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  buildInputs = [openssl wayland libxkbcommon libGL];
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:24]:
```suggestion
  buildInputs = [
    openssl
    wayland
    libxkbcommon
    libGL
  ];
```

File: pkgs/by-name/su/surfer/package.nix [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  # These libraries are dlopen'ed at runtime, but they won't be able to find anything in
+  # NixOS's path. So force them to be linked.
+  # This could alternatively be a wrapper which adds LD_LIBRARY_PATH.
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:28]: would `autoPatchelfHook` work here as well? might be a cleaner option imo

File: pkgs/by-name/su/surfer/package.nix [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "codespan-0.12.0" = "sha256-3F2006BR3hyhxcUTaQiOjzTEuRECKJKjIDyXonS/lrE=";
+      "egui_skia-0.5.0" = "sha256-dpkcIMPW+v742Ov18vjycLDwnn1JMsvbX6qdnuKOBC4=";
+      "tracing-tree-0.2.0" = "sha256-/JNeAKjAXmKPh0et8958yS7joORDbid9dhFB0VUAhZc=";
+    };
+  };
+
+  doCheck = false;
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:46]: a comment explaining why this is false would be nice

File: pkgs/by-name/su/surfer/package.nix [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+  meta = {
+    description = "An Extensible and Snappy Waveform Viewer";
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:49]:
```suggestion
    description = "Extensible and Snappy Waveform Viewer";
```
from the [contributing guidelines](https://github.com/NixOS/nixpkgs/tree/665394853091a08d679eb0c166900cb4f57a5bf5/pkgs#meta-attributes)

> - meta.description must:
>   - Be short, just one sentence.
>   - Be capitalized.
>   - ***Not start with the definite or an indefinite article.***

File: pkgs/by-name/su/surfer/package.nix [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    homepage = "http://surfer-project.org/";
+    license = lib.licenses.eupl12;
+    maintainers = with lib.maintainers; [hakan-demirli];
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:52]:
```suggestion
    maintainers = with lib.maintainers; [ hakan-demirli ];
```

File: pkgs/by-name/su/surfer/package.nix [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  libGL,
+}:
+rustPlatform.buildRustPackage rec {
+  pname = "surfer";
+  version = "v0.2.0";
+
+  src = fetchFromGitLab {
+    owner = "surfer-project";
+    repo = pname;
+    rev = version;
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:18]:
```suggestion
    rev = "v${version}";
```

File: maintainers/maintainer-list.nix
Comment (getchoo) [inline maintainers/maintainer-list.nix:1]: this should be split into a separate commit (prior to adding surfer itself)

**Review by getchoo — CHANGES_REQUESTED** [review]

**Comment by hakan-demirli** [comment]
Thanks for the review. My take aways are. I should have:
- Used `nixfmt-rfc-style` not `alejandra` for formatting.
- **Read** the docs not skim.


I will open another separate PR to add myself as a maintainer, the doc says it has a mandatory 1 week waiting period. Then I will fix this PR.

**Comment by getchoo** [comment]
you don't need to open another PR to add yourself as a maintainer; you only need a separate commit in this one

**Comment by hakan-demirli** [comment]
Result of `nixpkgs-review pr 324637` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>surfer</li>
  </ul>
</details>

File: pkgs/by-name/su/surfer/package.nix [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    autoPatchelfHook
+  ];
+
+  buildInputs = [
+    openssl
+    wayland
+    libxkbcommon
+    libGL
+    stdenv.cc.cc.libgcc
+  ];
```
Comment (dramforever) [inline pkgs/by-name/su/surfer/package.nix:36]:
```suggestion
    libX11
    libXcursor
    libXi
  ];
```

(and also the parameter list above)

These libraries are required for surfer to work on Xorg

**Review by Guanran928 — COMMENTED** [review]
Rewording the first commit would be nice,
image

File: pkgs/by-name/su/surfer/package.nix (hunk 10 above)
Comment (hakan-demirli) [reply pkgs/by-name/su/surfer/package.nix:42]: Surfer also supports MacOS platform however I dont have the hardware to test it. Is there a tool or nix command where I can test the derivation for cross-platform/cross-architecture?

I think I need to use `stdenv.hostPlatform.emulator` or `binfmt` but I just couldn't find the documentation about using these in nixpkgs testing. Or maybe I should just spin up a Qemu VM?

File: pkgs/by-name/su/surfer/package.nix [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+      "tracing-tree-0.2.0" = "sha256-/JNeAKjAXmKPh0et8958yS7joORDbid9dhFB0VUAhZc=";
+    };
+  };
+
+  # Avoid the network attempt from skia. See: https://github.com/cargo2nix/cargo2nix/issues/318
+  doCheck = false;
+
+  meta = {
+    description = "Extensible and Snappy Waveform Viewer";
+    homepage = "http://surfer-project.org/";
```
Comment (Guanran928) [inline pkgs/by-name/su/surfer/package.nix:65]: Sorry I didn't notice before, but a HTTPS link would be better.

Currently running `nixpkgs-review`

File: pkgs/by-name/su/surfer/package.nix [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    libXcursor
+    libXi
+  ];
+
+  runtimeDependencies = [
+    openssl
+    wayland
+    libxkbcommon
+    libGL
+  ];
```
Comment (dramforever) [inline pkgs/by-name/su/surfer/package.nix:49]: Very sorry for the extra trouble, but I misplaced the last comment about extra libraries. They were meant for `runtimeDependencies`. Like here: https://github.com/dramforever/nixpkgs/commit/72d029c70275e0ed2b41555787ad0dd7c4b28fca

Again sorry for the trouble 🙇

Comment (hakan-demirli) [reply pkgs/by-name/su/surfer/package.nix:49]: Nah, its my bad for not testing it on X11. I should probably create hydra like VM cluster, each having different wm and arch and run my CIs there.

Comment (dramforever) [reply pkgs/by-name/su/surfer/package.nix:49]: Well, the important thing is we got it working. Thanks!

File: pkgs/by-name/su/surfer/package.nix (hunk 11 above)
Comment (Guanran928) [reply pkgs/by-name/su/surfer/package.nix]: Sorry! was distracted by osu! and forgot to post it. :facepalm:

**Review by Guanran928 — APPROVED** [review]
Result of `nixpkgs-review pr 324637` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>surfer</li>
  </ul>
</details>

File: pkgs/by-name/su/surfer/package.nix [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    libxkbcommon
+    libGL
+    stdenv.cc.cc.libgcc
+  ];
+
+  runtimeDependencies = [
+    openssl
+    wayland
+    libxkbcommon
+    libGL
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:45]:
```suggestion
```
as these are already in buildInputs, they can be used autoPatchelfHook where necessary. i don't think we need to forcefully add these to the rpath of all binaries

File: pkgs/by-name/su/surfer/package.nix [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    libX11
+    libXcursor
+    libXi
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:48]:
```suggestion
    libX11
    libXcursor
    libXi
```
are these not detected by autoPatchelfHook?

File: pkgs/by-name/su/surfer/package.nix [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+      "tracing-tree-0.2.0" = "sha256-/JNeAKjAXmKPh0et8958yS7joORDbid9dhFB0VUAhZc=";
+    };
+  };
+
+  # Avoid the network attempt from skia. See: https://github.com/cargo2nix/cargo2nix/issues/318
+  doCheck = false;
+
+  meta = {
+    description = "Extensible and Snappy Waveform Viewer";
+    homepage = "https://surfer-project.org/";
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:65]:
```suggestion
    homepage = "https://surfer-project.org/";
    changelog = "https://gitlab.com/surfer-project/surfer/-/releases/v${version}";
```

**Review by getchoo — CHANGES_REQUESTED** [review]

File: pkgs/by-name/su/surfer/package.nix (hunk 13 above)
Comment (hakan-demirli) [reply pkgs/by-name/su/surfer/package.nix:45]: When I remove `wayland` and `libxkbcommon` from runtimeDependencies I get
```
event_loop/mod.rs", error: WaylandError(Connection(NoWaylandLib))
```
Alternative is to set
 `          LD_LIBRARY_PATH = "${lib.makeLibraryPath buildInputs}";`
 which is worse imo.

File: pkgs/by-name/su/surfer/package.nix (hunk 14 above)
Comment (hakan-demirli) [reply pkgs/by-name/su/surfer/package.nix:48]: When I remove them and run surfer on Ubuntu using NixGL I get:
```
Message:  called `Result::unwrap()` on an `Err` value: WinitEventLoop(NotSupported(NotSupportedError))
```

It works when I add the libs back.

File: pkgs/by-name/su/surfer/package.nix (hunk 13 above)
Comment (dramforever) [reply pkgs/by-name/su/surfer/package.nix:45]: winit uses `dlopen` to open X and Wayland library, for better or for worse. Therefore, `runtimeDependencies` is required:

> You can also specify a `runtimeDependencies` variable which lists dependencies to be unconditionally added to rpath of all executables. This is useful for programs that use `dlopen(3)` to load libraries at runtime.

https://nixos.org/manual/nixpkgs/stable/#setup-hook-autopatchelfhook

File: pkgs/by-name/su/surfer/package.nix (hunk 14 above)
Comment (dramforever) [reply pkgs/by-name/su/surfer/package.nix:48]: This is also required because of dlopen. See https://github.com/NixOS/nixpkgs/pull/324637#discussion_r1667592609. Maybe add a comment?

File: pkgs/by-name/su/surfer/package.nix (hunk 13 above)
Comment (getchoo) [reply pkgs/by-name/su/surfer/package.nix:45]: > winit uses dlopen to open X and Wayland libraries and libGL, for better or for worse.

yeah, this is what i assumed when hakan reported the error. you can disregard this

though i do wonder if we need them in builldInputs as well if that's the case? maybe since they're just using dlopen, we don't need to resolve with the pkg-config

Comment (hakan-demirli) [reply pkgs/by-name/su/surfer/package.nix:46]: @getchoo What can I do to resolve this conversation? I have added a comment explaining the dlopen situation. Should I also try to remove pkg-config dependency? I am not familiar with details of linking/finding libs on nix so a more direct request would be helpful.

File: pkgs/by-name/su/surfer/package.nix [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  nativeBuildInputs = [
+    pkg-config
+    autoPatchelfHook
+  ];
+
+  buildInputs = [
+    openssl
+    wayland
+    libxkbcommon
+    libGL
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:37]:
```suggestion
```

these aren't used by pkg-config and aren't required for compilation

File: pkgs/by-name/su/surfer/package.nix [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    stdenv.cc.cc.libgcc
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:38]:
```suggestion
    stdenv.cc.cc.lib
```

File: pkgs/by-name/su/surfer/package.nix [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  ];
+
+  # wayland and x11 libs required at runtime since winit uses dlopen
+  runtimeDependencies = [
+    openssl
```
Comment (getchoo) [inline pkgs/by-name/su/surfer/package.nix:43]:
```suggestion
```
this is already dynamically linked. patching all rpaths with this is unnecessary

**Review by getchoo — CHANGES_REQUESTED** [review]

File: pkgs/by-name/su/surfer/package.nix (hunk 13 above)
Comment (getchoo) [reply pkgs/by-name/su/surfer/package.nix:46]: this can be resolved now. please see my newer comments

**Comment by kacper-uminski** [comment]
Contributor to surfer here. It would be most appreciated to have this merged. :) Is there anything left that needs to be done which I could help with?

**Comment by bgamari** [comment]
While this appears to build, any attempt to use functions involving the file picker fail with `[ERROR] pick_file error No such file or directory (os error 2)`.

**Comment by bgamari** [comment]
> While this appears to build and is largely functional, any attempt to use functions involving the file picker fail with `[ERROR] pick_file error No such file or directory (os error 2)`.

The reason for this appears to be that the package somehow depends upon `nixpkgs#gnome3.zenity` for file selection. `nix shell nixpkgs#gnome3.zenity github:hakan-demirli/nixpkgs/surfer#surfer -c surfer` works as expected.

**Comment by Aleksanaa** [comment]
> The reason for this appears to be that the package somehow depends upon `nixpkgs#gnome3.zenity` for file selection. `nix shell nixpkgs#gnome3.zenity github:hakan-demirli/nixpkgs/surfer#surfer -c surfer` works as expected.

This is because it's using https://github.com/PolyMeilex/rfd/blob/e0e725ec9a426acf7c93d9877c41e230fcee3fac/src/backend/linux/zenity.rs to open file picker. We can just wrap it with zenity.

File: pkgs/by-name/su/surfer/package.nix (hunk 1 above)
Comment (Aleksanaa) [reply pkgs/by-name/su/surfer/package.nix]: Remember to edit commit message accordingly
